### PR TITLE
Add new produce API call, returning the assigned offset 

### PR DESCRIPTION
**Pull request by aronisstav**
Kafka responses for produce requests that are acknowledged include the offset that was assigned to the message. This PR adds an API call for retrieving this offset. 

It also includes a number of small fixes.

Feedback is very welcome. :-)

### CODE REVIEW

**Comment by aronisstav** [comment]
This Travis job needs to be restarted: https://travis-ci.org/klarna/brod/jobs/397265533

**Comment by zmstone** [comment]
The offset in produce response is the base offset of a produced batch.
Messages from multiple `brod:produce` calls may get batched in to one produce request,
hence replying the same offset for all calls may not be optimal.
we can try to calculate offsets for each message.

Please note that a producer may send batches themselves as
```
brod:produce(Client, Topic, Partition, _Key =<<>>, _Value = [{K1, V1}, {K2, V2}]).
% where V1 and V2 can also be nested batches
```
i.e. we'll need to `Offset = LastOffset +1` for each of the inner batches while calculating offsets.

File: include/brod.hrl [hunk 1]
```diff
@@ -44,7 +44,10 @@
                        , ref    :: reference()
                        }).
 
+-define(BROD_PRODUCE_UNKNOWN_OFFSET, -1).
+
 -record(brod_produce_reply, { call_ref :: brod:call_ref()
+                            , offset   :: brod:offset()
```
Comment (zmstone) [inline include/brod.hrl]: maybe call it `base_offset` as it is in the produce response.

Comment (aronisstav) [reply include/brod.hrl]: Good point!

File: src/brod_utils.erl [hunk 2]
```diff
@@ -497,6 +498,26 @@ bytes(?TKV(T, K, V)) when is_integer(T) -> 8 + bytes(K) + bytes(V);
...
+
+%% @private Foldl kv-list.
+foldl_kvlist(_Fun, Acc, []) -> Acc;
+foldl_kvlist(Fun, Acc, [Msg | Rest]) ->
+  NewAcc = case nested(Msg) of
+             false -> Fun(Msg, Acc);
+             Nested -> foldl_kvlist(Fun, Acc, Nested)
+           end,
+  foldl_kvlist(Fun, NewAcc, Rest).
+
```
Comment (aronisstav) [inline src/brod_utils.erl:520]: I copied this code from `kafka_protocol/src/kpro.erl`.

File: src/brod_producer_buffer.erl [hunk 3]
```diff
@@ -137,14 +141,20 @@ maybe_send(#buf{} = Buf, SockPid, Vsn) ->
...
+  ack(Buf, CorrId, ?BROD_PRODUCE_UNKNOWN_OFFSET).
+
+-spec ack(buf(), corr_id(), offset()) ->
+             {ok, buf()} | {error, none | corr_id()}.
 ack(#buf{ onwire_count = OnWireCount
         , onwire       = [{CorrId, Reqs} | Rest]
-        } = Buf, CorrId) ->
-  ok = lists:foreach(fun reply_acked/1, Reqs),
+        } = Buf, CorrId, BaseOffset) ->
+  Fold = fun(R, Offset) -> reply_acked(R, Offset) end,
```
Comment (zmstone) [inline src/brod_producer_buffer.erl]: `Fold = fun reply_acked/2`

**Comment by zmstone** [comment]
very nice !

File: src/brod_producer_buffer.erl (hunk 3 above)
Comment (aronisstav) [reply src/brod_producer_buffer.erl]: True!

**Review by viktoriafordos — COMMENTED** [review]
~~There are non-backward compatible changes in this PR. 
For instance, sync_produce_request will return {ok, offset()} instead of ok. This will require all client to change their code if they want to upgrade brod. I think this can be avoided. However, I don't know what is the contribution policy in general for the brod library.
Are we fine with being non-backward compatible @zmstone ?~~

@aronisstav pointed out that the API module is brod. The signature of existing functions in brod has not been changed.

Another point is that pending async produces will not be acknowledged even though they were successful because the brod_produce_reply record has changed and is used in a receive. So upgrading the application may yield to false error.

**Review by zmstone — APPROVED** [review]

**Review by zmstone — CHANGES_REQUESTED** [review]
Sorry, missed two things:
1. Please bump version to 3.5.0 in both .app.src and Makefile.
2. Please update changelog.md to describe this new feature (and hint that upgrading to this version will require restarts of all producers).

**Comment by aronisstav** [comment]
@zmstone Done requested changes.

Regarding @viktoriafordos comment, the changelog makes the potential issues on upgrade explicit.

**Review by zmstone — APPROVED** [review]

File: test/brod_producer_SUITE.erl [hunk 4]
```diff
@@ -34,16 +34,20 @@
...
         , t_producer_topic_not_found/1
         , t_producer_partition_not_found/1
         , t_produce_partitioner/1
         , t_produce_batch/1
+        , t_produce_buffered_offset/1
         ]).
 
 -include_lib("common_test/include/ct.hrl").
 -include_lib("eunit/include/eunit.hrl").
+-include("brod.hrl").
```
Comment (aronisstav) [inline test/brod_producer_SUITE.erl]: I noticed today that this is redundant. Should I remove it?

File: test/brod_producer_SUITE.erl [hunk 5]
```diff
@@ -57,7 +61,7 @@ subscriber_loop(Client, TesterPid) ->
       #kafka_message_set{ messages = Messages
                         , partition = Partition} = KMS,
       lists:foreach(fun(#kafka_message{offset = Offset, key = K, value = V}) ->
-                      TesterPid ! {Partition, K, V},
+                      TesterPid ! {Partition, Offset, K, V},
                       ok = brod:consume_ack(ConsumerPid, Offset)
                     end, Messages),
       subscriber_loop(Client, TesterPid);
```
Comment (aronisstav) [inline test/brod_producer_SUITE.erl]: `Client` is unused in this function (only going to the tail call). Should I remove it?

File: test/brod_producer_SUITE.erl (hunk 4 above)
Comment (zmstone) [reply test/brod_producer_SUITE.erl]: yes please.

File: test/brod_producer_SUITE.erl (hunk 5 above)
Comment (zmstone) [reply test/brod_producer_SUITE.erl]: yes please.

**Review by zmstone — APPROVED** [review]

**Review by k32 — APPROVED** [review]

**Pull request merged**